pkg/controller/servicecatalog/provisionedproduct/setup.go

```diff
@@ -0,0 +1,367 @@
+	ds.Status.AtProvider.Status = resp.ProvisionedProductDetail.Status
+	ds.Status.AtProvider.StatusMessage = resp.ProvisionedProductDetail.StatusMessage
+	ds.Status.AtProvider.ProvisionedProductType = resp.ProvisionedProductDetail.Type
+	ds.Status.AtProvider.RecordType = describeRecordOutput.RecordDetail.RecordType
+	ds.Status.AtProvider.LastPathID = describeRecordOutput.RecordDetail.PathId
+	ds.Status.AtProvider.LastProductID = describeRecordOutput.RecordDetail.ProductId
+	ds.Status.AtProvider.LastProvisioningArtifactID = describeRecordOutput.RecordDetail.ProvisioningArtifactId
+	ds.Status.AtProvider.LastProvisioningParameters = ds.Spec.ForProvider.ProvisioningParameters
+
+	return obs, nil
+}
+
+func preDelete(_ context.Context, ds *svcapitypes.ProvisionedProduct, input *svcsdk.TerminateProvisionedProductInput) (bool, error) {
+	if ptr.Deref(ds.Status.AtProvider.Status, "") == string(svcapitypes.ProvisionedProductStatus_SDK_UNDER_CHANGE) {
+		return true, nil
+	}
+	input.TerminateToken = aws.String(genIdempotencyToken())
+	if ds.GetName() == meta.GetExternalName(ds) {
+		input.ProvisionedProductName = aws.String(meta.GetExternalName(ds))
+	} else {
+		input.ProvisionedProductId = aws.String(meta.GetExternalName(ds))
+	}
+	return false, nil
+}
+
+func setConditions(describeRecordOutput *svcsdk.DescribeRecordOutput, resp *svcsdk.DescribeProvisionedProductOutput, cr *svcapitypes.ProvisionedProduct) {
+	ppStatus := aws.StringValue(resp.ProvisionedProductDetail.Status)
+	switch {
+	case ppStatus == string(svcapitypes.ProvisionedProductStatus_SDK_AVAILABLE):
+		cr.SetConditions(xpv1.Available())
+	case ppStatus == string(svcapitypes.ProvisionedProductStatus_SDK_UNDER_CHANGE):
+		recordType := ptr.Deref(describeRecordOutput.RecordDetail.RecordType, "UPDATE_PROVISIONED_PRODUCT")
+		switch {
+		case recordType == "PROVISION_PRODUCT":
+			cr.SetConditions(xpv1.Creating())
+		case recordType == "UPDATE_PROVISIONED_PRODUCT":
+			cr.SetConditions(xpv1.Unavailable().WithMessage(msgProvisionedProductStatusSdkUnderChange))
+		case recordType == "TERMINATE_PROVISIONED_PRODUCT":
+			cr.SetConditions(xpv1.Deleting())
+		}
+	case ppStatus == string(svcapitypes.ProvisionedProductStatus_SDK_PLAN_IN_PROGRESS):
+		cr.SetConditions(xpv1.Unavailable().WithMessage(msgProvisionedProductStatusSdkPlanInProgress))
+	case ppStatus == string(svcapitypes.ProvisionedProductStatus_SDK_ERROR):
+		cr.SetConditions(xpv1.Unavailable().WithMessage(msgProvisionedProductStatusSdkError))
+	case ppStatus == string(svcapitypes.ProvisionedProductStatus_SDK_TAINTED):
+		cr.SetConditions(xpv1.Unavailable().WithMessage(msgProvisionedProductStatusSdkTainted))
+	}
+}
+
+func (c *custom) provisioningParamsAreChanged(ctx context.Context, ds *svcapitypes.ProvisionedProduct) (bool, error) {
+	nn := types.NamespacedName{
+		Name: ds.GetName(),
+	}
+	xr := svcapitypes.ProvisionedProduct{}
+	err := c.kube.Get(ctx, nn, &xr)
+	if err != nil {
+		return false, err
+	}
```

Comment:
I don't think we need to call K8s a second time as we already have the latest version of the MR.

What should happen to compare MRs with the external state:

1. Observe external resource
2. LateInialize MR fields with external values if unset (to detect late initialization, you can create a deep copy)
3. Compare the late initialized MR with the observed state. There are multiple ways to do this: One would be to compare all fields manually or generate a spec from the external state and compare it with `cmp.Diff`.

Reply:
If I understood you correctly, but I'm not sure, it won't work.

The context and a problem are that:
external resource(cloudformation stack which can be found by  CloudformationStackARN from a provisioned product outputs) keeps all provisioning params every time, even if some of them are not specified when a product was ran, some times value for not specified params is empty string, sometimes it is a default value, depends on product.  
E.g. we have a product with 10 params, but a user specified only two of them, in this case CloudformationStackARN will have all 10 params, 2 of them set by an user and others have default or empty values.

In this case it is possible to detect if some of parameters where deleted only if a user specified all of them, which is not necessary and usually users specified only what a part of parameters. Ss mentioned before, external resource has all parameters and if MR has only 2/10 of params, how can we be sure it was not have 3/10 before, comparison with exteranl shows nothing. 
That's why I added atProvider.status.lastProvisiongParams which are ProvisioningParams from the previous reconciliation loop, and the controller use them for comparison with current MR params.

>or generate a spec from the external state and compare it with cmp.Diff
I used cmp.Equal because it allows to sort array before comparison, the idea is do not run update if only sequence of params are changred

Reply:
You already have the latest status of the MR during reconcile. Because it is overwritten during observe (I guess thats why you reretrieved it from the cluster) you can store a copy of it in `custom` during `preObserve`.

> I used cmp.Equal because it allows to sort the arrays before comparison, the idea is do not run update if only a sequence of params is changred

`isUpToDate` allows you to return a diff that is printed in the logs and might help users to determine issues during reconciles. It might be a good idea to use `cmp.Diff` and return its result.

Reply:
>You already have the latest status of the MR during reconcile

Exactly, *svcapitypes.ProvisionedProduct represent MR during reconcile, but for detection of removed paramers we should compare provisioning params(Spec.ForProvider.ProvisioningParameters) with params from previous reconcile(Status.AtProvider.LastProvisioningParameters) but *svcapitypes.ProvisionedProduct doesn't contian this data before we set it postObserve(this data will be used for comparison with provisioning params in the next reconciliation loop)

Reply:
By "latest status of the MR" I meant the params from the previous reconcile. There is no need to call `kube.Get` a second time because you can just store `cr.DeepCopy()` in `custom.cache` during `preObserve` before `cr` gets updated with the external values.

Reply:
Sorry, I didn't understand the idea the first time(and second too : ) ), now I got it. Thanks!
Fixed this. I also tried cmp.Diff and it is too humanreadable for one line logging, \n and spaces looks nor realy nice. So I   left cmp.Equal.

Reply:
> I also tried cmp.Diff and it is too humanreadable for one line logging

Yeah, it is not the ideal thing, but better than nothing. However, `cmp.Diff` helps a lot to debug reconcile bugs so I would prefer it instead of `cmp.Equal`.

Reply:
Done

Reply:
@MisterMX is it okay, could you check?